File: src2/camera.lua
local M

M = class {
  init = function(self)
    self.x = 0
    self.y = 0
    self.width = push:getWidth() 
    self.height = push:getHeight()
    self.transform = love.math.newTransform()
  end,
  start = function(self)
    love.graphics.push()
    self.transform:reset()
    self.transform:translate(self.width/2, self.height/2)
    self.transform:translate(-self.x, -self.y)
    love.graphics.applyTransform(self.transform)
  end,
  finish = function(self)
    love.graphics.pop()
  end,
  toWorld = function(self, x, y)
    return self.transform:inverseTransformPoint(x, y)
  end
}

return M 

File: tssrc/camera.ts
import { Transform } from 'love.math'
import * as push from './libs/push'

export class Camera {
  x: number = 0
  y: number = 0
  width: number = 0
  height: number = 0
  transform: Transform = love.math.newTransform()

  constructor() {
    this.width = push.getWidth()
    this.height = push.getHeight()
  }

  start() {
    love.graphics.push()
    this.transform.reset()
    this.transform.translate(this.width / 2, this.height / 2)
    this.transform.translate(-this.x, -this.y)
    love.graphics.applyTransform(this.transform)
  }

  finish() {
    love.graphics.pop()
  }

  toWorld(x: number, y: number) {
    return this.transform.inverseTransformPoint(x, y)
  }
}
